odl/test/space/pspace_test.py

```diff
@@ -997,50 +997,66 @@ def test_real_imag_and_conj():
 
 
 def test_real_setter_product_space(space, newpart):
-    """Verify that the setter for the real part of an element works."""
+    """Verify that the setter for the real part of an element works.
+    What setting the real part means depends on the inputs; we perform a
+    recursive deconstruction to cover the possible cases."""
+
+    def verify_result(x, expected_result, recursion_limit=4):
+        if recursion_limit <= 0:
+            return False
+        try:
+            # Catch scalar argument
+            iter(expected_result)
+        except TypeError:
+            return verify_result(x, expected_result * space.one(),
+                                 recursion_limit - 1)
+        if expected_result in space:
+            return all_equal(x.real, expected_result.real)
+        elif all_equal(x.real, expected_result):
+            return True
+        elif space.is_power_space:
+            return verify_result(x, [expected_result for _ in space],
+                                 recursion_limit - 1)
+
```

Comment:
Why recursion limit is 4? Should not it be 2?

Reply:
Well, the `recursion_limit` is never supposed to be actually reached, it's only there to ensure that the deconstruction doesn't end up in some infinite recursion scenario (which would result in a very strange stack-overflow failure of the test suite, if not just a hung test driver).

In practice `recursion_limit=2` would also be enough here, yes, but it would be conceivable to unwrap nested product spaces, in which case that would not suffice. Maybe some larger number would make it clearer what's the intent?

Reply:
Perhaps you can leave a comment saying that the limit us not supposed to be reached?

Reply:
[done](https://github.com/odlgroup/odl/pull/1640/commits/67e5eee133b122c3f666e3c8bd16f85765bd8357)